girder/models/file.py

```diff
@@ -94,18 +95,35 @@ def download(self, file, offset=0, headers=True, endByte=None,
         :type extraParameters: str or None
         """
         if file.get('assetstoreId'):
-            return self.getAssetstoreAdapter(file).downloadFile(
-                file, offset=offset, headers=headers, endByte=endByte,
-                contentDisposition=contentDisposition,
-                extraParameters=extraParameters)
+            try:
+                fileDownload = self.getAssetstoreAdapter(file).downloadFile(
+                    file, offset=offset, headers=headers, endByte=endByte,
+                    contentDisposition=contentDisposition,
+                    extraParameters=extraParameters)
+                if inspect.isgeneratorfunction(fileDownload) or inspect.isgenerator(fileDownload):
+                    def downloadGenerator():
+                        for data in fileDownload():
+                            yield data
+                        if endByte is None or endByte >= file['size']:
+                            events.trigger('download.complete', info={'file': file,
+                                                                      'redirect': False})
+                    return downloadGenerator
+                else:
+                    return fileDownload
+            except cherrypy.HTTPRedirect as e:
+                events.trigger('download.complete', info={'file': file, 'redirect': True})
+                raise e
```

Comment:
In cases where we want to add some additional behavior to an exception propagation, but don't want to actually alter the exception or its propagation, it's better to use [a plain `raise` statement](https://docs.python.org/2/reference/simple_stmts.html#raise), without a variable. That will preserve the exception's built-in stack trace all the way down to the original place it was raised.

So, this can be simply:
```
except cherrypy.HTTPRedirect:
    events.trigger('download.complete', info={'file': file, 'redirect': True})
    raise
```
since we no longer need `e`.